File: PythonAPI/carla/agents/tools/misc.py
#!/usr/bin/env python

# Copyright (c) 2018 Intel Labs.
# authors: German Ros (german.ros@intel.com)
#
# This work is licensed under the terms of the MIT license.
# For a copy, see <https://opensource.org/licenses/MIT>.

"""Module with auxiliary functions."""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

import numpy as np

import carla

if TYPE_CHECKING:
    from collections.abc import Iterable

    from carla import Location, TrafficLight, Transform, Vehicle, Waypoint, World


def draw_waypoints(world: World, waypoints: Iterable[Waypoint], z: float = 0.5) -> None:
    """Draw a list of waypoints at a certain height given in z.

    Args:
        world: carla.World object.
        waypoints: Iterable container with the waypoints to draw.
        z: Height in meters.
    """
    for wpt in waypoints:
        wpt_t = wpt.transform
        begin = wpt_t.location + carla.Location(z=z)
        angle = math.radians(wpt_t.rotation.yaw)
        end = begin + carla.Location(x=math.cos(angle), y=math.sin(angle))
        world.debug.draw_arrow(begin, end, arrow_size=0.3, life_time=1.0)


def get_speed(vehicle: Vehicle) -> float:
    """Compute speed of a vehicle in Km/h.

    Args:
        vehicle: The vehicle for which speed is calculated.

    Returns:
        Speed as a float in Km/h.
    """
    vel = vehicle.get_velocity()
    return 3.6 * math.sqrt(vel.x**2 + vel.y**2 + vel.z**2)


def get_trafficlight_trigger_location(traffic_light: TrafficLight) -> Location:
    """Calculate the location of the trigger volume of the traffic light.

    Args:
        traffic_light: The traffic light to get trigger location for.

    Returns:
        Location of the trigger volume.
    """

    def rotate_point(point: carla.Vector3D, radians: float) -> carla.Vector3D:
        """Rotate a given point by a given angle."""
        rotated_x = math.cos(radians) * point.x - math.sin(radians) * point.y
        rotated_y = math.sin(radians) * point.x - math.cos(radians) * point.y
        return carla.Vector3D(rotated_x, rotated_y, point.z)

    base_transform = traffic_light.get_transform()
    base_rot = base_transform.rotation.yaw
    area_loc = base_transform.transform(traffic_light.trigger_volume.location)
    area_ext = traffic_light.trigger_volume.extent

    point = rotate_point(carla.Vector3D(0, 0, area_ext.z), math.radians(base_rot))
    point_location = area_loc + carla.Location(x=point.x, y=point.y)

    return carla.Location(point_location.x, point_location.y, point_location.z)


def is_within_distance(
    target_transform: Transform,
    reference_transform: Transform,
    max_distance: float,
    angle_interval: tuple[float, float] | None = None,
) -> bool:
    """Check if a location is within a certain distance from a reference object.

    By using 'angle_interval', the angle between the location and reference transform
    will also be taken into account, being 0 a location in front and 180, one behind.

    Args:
        target_transform: Transform of the target object.
        reference_transform: Transform of the reference object.
        max_distance: Maximum allowed distance.
        angle_interval: Only locations between [min, max] angles will be considered.

    Returns:
        True if within distance and angle constraints.
    """
    target_vector = np.array(
        [
            target_transform.location.x - reference_transform.location.x,
            target_transform.location.y - reference_transform.location.y,
        ]
    )
    norm_target = np.linalg.norm(target_vector)

    if norm_target < 0.001:
        return True

    if norm_target > max_distance:
        return False

    if not angle_interval:
        return True

    min_angle = angle_interval[0]
    max_angle = angle_interval[1]

    fwd = reference_transform.get_forward_vector()
    forward_vector = np.array([fwd.x, fwd.y])
    angle = math.degrees(math.acos(np.clip(np.dot(forward_vector, target_vector) / norm_target, -1.0, 1.0)))

    return min_angle < angle < max_angle


def compute_magnitude_angle(
    target_location: Location,
    current_location: Location,
    orientation: float,
) -> tuple[float, float]:
    """Compute relative angle and distance between a target_location and a current_location.

    Args:
        target_location: Location of the target object.
        current_location: Location of the reference object.
        orientation: Orientation of the reference object in degrees.

    Returns:
        Tuple of (distance, angle) between the objects.
    """
    target_vector = np.array([target_location.x - current_location.x, target_location.y - current_location.y])
    norm_target = np.linalg.norm(target_vector)

    forward_vector = np.array([math.cos(math.radians(orientation)), math.sin(math.radians(orientation))])
    d_angle = math.degrees(math.acos(np.clip(np.dot(forward_vector, target_vector) / norm_target, -1.0, 1.0)))

    return norm_target, d_angle


def distance_vehicle(waypoint: Waypoint, vehicle_transform: Transform) -> float:
    """Return the 2D distance from a waypoint to a vehicle.

    Args:
        waypoint: Actual waypoint.
        vehicle_transform: Transform of the target vehicle.

    Returns:
        2D distance in meters.
    """
    loc = vehicle_transform.location
    x = waypoint.transform.location.x - loc.x
    y = waypoint.transform.location.y - loc.y

    return math.sqrt(x * x + y * y)


def vector(location_1: Location, location_2: Location) -> list[float]:
    """Return the unit vector from location_1 to location_2.

    Args:
        location_1: Starting location.
        location_2: Ending location.

    Returns:
        Unit vector as [x, y, z].
    """
    x = location_2.x - location_1.x
    y = location_2.y - location_1.y
    z = location_2.z - location_1.z
    norm = np.linalg.norm([x, y, z]) + np.finfo(float).eps

    return [x / norm, y / norm, z / norm]


def compute_distance(location_1: Location, location_2: Location) -> float:
    """Euclidean distance between 3D points.

    Args:
        location_1: First location.
        location_2: Second location.

    Returns:
        Distance in meters.
    """
    x = location_2.x - location_1.x
    y = location_2.y - location_1.y
    z = location_2.z - location_1.z
    return np.linalg.norm([x, y, z]) + np.finfo(float).eps


def positive(num: float) -> float:
    """Return the given number if positive, else 0.

    Args:
        num: Value to check.

    Returns:
        num if positive, else 0.0.
    """
    return max(0.0, num)

File: LibCarla/source/carla/agents/navigation/Misc.cpp
// Copyright (c) 2018-2020 CVC.
//
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT>.

#include "carla/agents/navigation/Misc.h"

#include "carla/client/Actor.h"
#include "carla/client/DebugHelper.h"
#include "carla/client/TrafficLight.h"
#include "carla/client/Waypoint.h"
#include "carla/client/World.h"
#include "carla/geom/BoundingBox.h"
#include "carla/geom/Location.h"
#include "carla/geom/Math.h"
#include "carla/geom/Transform.h"
#include "carla/geom/Vector3D.h"

#include <algorithm>
#include <cmath>
#include <limits>

namespace carla {
namespace agents {
namespace navigation {

namespace {

constexpr float kKmhPerMs = 3.6f;

// Replicates numpy's `np.finfo(float).eps` regularisation used in the Python
// helpers so we don't divide by zero on coincident points.
constexpr float kNormEps = std::numeric_limits<float>::epsilon();

} // namespace

float GetSpeed(SharedPtr<client::Actor> actor) {
  if (!actor) {
    return 0.0f;
  }
  const auto vel = actor->GetVelocity();
  return kKmhPerMs * std::sqrt(vel.x * vel.x + vel.y * vel.y + vel.z * vel.z);
}

geom::Location GetTrafficLightTriggerLocation(
    SharedPtr<client::TrafficLight> traffic_light) {
  // Local helper that mirrors the lambda in misc.py. The Python code rotates
  // the (0, 0, extent.z) point by the traffic light's yaw angle and adds it
  // back onto the trigger volume's transformed location.
  auto rotate_point = [](const geom::Vector3D &point, float radians) {
    const float rotated_x =
        std::cos(radians) * point.x - std::sin(radians) * point.y;
    const float rotated_y =
        std::sin(radians) * point.x - std::cos(radians) * point.y;
    return geom::Vector3D(rotated_x, rotated_y, point.z);
  };

  const geom::Transform base_transform = traffic_light->GetTransform();
  const float base_rot = base_transform.rotation.yaw;
  const geom::BoundingBox &trigger = traffic_light->GetTriggerVolume();

  // Transform the trigger volume's local centre into world space.
  geom::Vector3D area_loc(trigger.location.x, trigger.location.y,
                          trigger.location.z);
  base_transform.TransformPoint(area_loc);

  const geom::Vector3D &area_ext = trigger.extent;

  const geom::Vector3D point = rotate_point(
      geom::Vector3D(0.0f, 0.0f, area_ext.z),
      geom::Math::ToRadians(base_rot));

  return geom::Location(area_loc.x + point.x,
                        area_loc.y + point.y,
                        area_loc.z);
}

bool IsWithinDistance(
    const geom::Transform &target,
    const geom::Transform &reference,
    float max_distance,
    std::optional<std::pair<float, float>> angle_interval) {
  const float dx = target.location.x - reference.location.x;
  const float dy = target.location.y - reference.location.y;
  const float norm_target = std::sqrt(dx * dx + dy * dy);

  // Match the early-out behaviour from misc.py.
  if (norm_target < 0.001f) {
    return true;
  }
  if (norm_target > max_distance) {
    return false;
  }
  if (!angle_interval.has_value()) {
    return true;
  }

  const float min_angle = angle_interval->first;
  const float max_angle = angle_interval->second;

  const auto fwd = reference.GetForwardVector();
  const float dot = (fwd.x * dx + fwd.y * dy) / norm_target;
  const float clamped = std::clamp(dot, -1.0f, 1.0f);
  const float angle_deg = geom::Math::ToDegrees(std::acos(clamped));

  return min_angle < angle_deg && angle_deg < max_angle;
}

std::pair<float, float> ComputeMagnitudeAngle(
    const geom::Location &target,
    const geom::Location &current,
    float orientation_deg) {
  const float dx = target.x - current.x;
  const float dy = target.y - current.y;
  const float norm_target = std::sqrt(dx * dx + dy * dy);

  const float orientation_rad = geom::Math::ToRadians(orientation_deg);
  const float fwd_x = std::cos(orientation_rad);
  const float fwd_y = std::sin(orientation_rad);

  // Avoid a divide-by-zero when target and current coincide.
  const float denom = norm_target > 0.0f ? norm_target : kNormEps;
  const float dot = (fwd_x * dx + fwd_y * dy) / denom;
  const float clamped = std::clamp(dot, -1.0f, 1.0f);
  const float angle_deg = geom::Math::ToDegrees(std::acos(clamped));

  return {norm_target, angle_deg};
}

float DistanceVehicle(
    SharedPtr<client::Waypoint> waypoint,
    const geom::Transform &vehicle_transform) {
  const auto &wp_loc = waypoint->GetTransform().location;
  const float dx = wp_loc.x - vehicle_transform.location.x;
  const float dy = wp_loc.y - vehicle_transform.location.y;
  return std::sqrt(dx * dx + dy * dy);
}

std::array<float, 3> UnitVector(
    const geom::Location &a,
    const geom::Location &b) {
  const float x = b.x - a.x;
  const float y = b.y - a.y;
  const float z = b.z - a.z;
  // Match Python's `np.linalg.norm(...) + np.finfo(float).eps`.
  const float norm = std::sqrt(x * x + y * y + z * z) + kNormEps;
  return {x / norm, y / norm, z / norm};
}

float ComputeDistance(
    const geom::Location &a,
    const geom::Location &b) {
  const float x = b.x - a.x;
  const float y = b.y - a.y;
  const float z = b.z - a.z;
  return std::sqrt(x * x + y * y + z * z) + kNormEps;
}

void DrawWaypoints(
    SharedPtr<client::World> world,
    const std::vector<SharedPtr<client::Waypoint>> &waypoints,
    float z) {
  if (!world) {
    return;
  }
  auto debug = world->MakeDebugHelper();
  for (const auto &wpt : waypoints) {
    if (!wpt) continue;
    const auto wpt_t = wpt->GetTransform();
    const geom::Location begin(
        wpt_t.location.x,
        wpt_t.location.y,
        wpt_t.location.z + z);
    const float angle = geom::Math::ToRadians(wpt_t.rotation.yaw);
    const geom::Location end(
        begin.x + std::cos(angle),
        begin.y + std::sin(angle),
        begin.z);
    debug.DrawArrow(
        begin,
        end,
        /*thickness=*/0.1f,
        /*arrow_size=*/0.3f,
        client::DebugHelper::Color{255u, 0u, 0u},
        /*life_time=*/1.0f);
  }
}

} // namespace navigation
} // namespace agents
} // namespace carla
